#### Nama : Rahmatul Idami
#### NPM : 2108017010071

#### Contoh pada artikel kedua menggunakan tensorflow pada python environment

Datasetnya yang digunakan yaitu dataset data_bank_churn yang di download [disini](https://www.megabagus.id/download/data-ann/)

In [1]:
# Mengimpor library yang diperlukan
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

In [3]:
# Mengimpor dataset
dataset = pd.read_csv('Dataset/Data_bank_churn.csv')
dataset

RowNumber  CustomerId    Surname  CreditScore Geography  Gender  Age  \
0             1    15634602   Hargrave          619    France  Female   42   
1             2    15647311       Hill          608     Spain  Female   41   
2             3    15619304       Onio          502    France  Female   42   
3             4    15701354       Boni          699    France  Female   39   
4             5    15737888   Mitchell          850     Spain  Female   43   
...         ...         ...        ...          ...       ...     ...  ...   
9995       9996    15606229   Obijiaku          771    France    Male   39   
9996       9997    15569892  Johnstone          516    France    Male   35   
9997       9998    15584532        Liu          709    France  Female   36   
9998       9999    15682355  Sabbatini          772   Germany    Male   42   
9999      10000    15628319     Walker          792    France  Female   28   

      Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0          2       0.00              1          1               1   
1          1   83807.86              1          0               1   
2          8  159660.80              3          1               0   
3          1       0.00              2          0               0   
4          2  125510.82              1          1               1   
...      ...        ...            ...        ...             ...   
9995       5       0.00              2          1               0   
9996      10   57369.61              1          1               1   
9997       7       0.00              1          0               1   
9998       3   75075.31              2          1               0   
9999       4  130142.79              1          1               0   

      EstimatedSalary  Exited  
0           101348.88       1  
1           112542.58       0  
2           113931.57       1  
3            93826.63       0  
4            79084.10       0  
...               ...     ...  
9995         96270.64       0  
9996        101699.77       0  
9997         42085.58       1  
9998         92888.52       1  
9999         38190.78       0  

[10000 rows x 14 columns]

In [4]:
X = dataset.iloc[:, 3:13].values
y = dataset.iloc[:, 13].values

- Pertama dilakukan slicing untuk variabel independen X. dibutuhkan kolom ke 3-12. 
- Kemudian dilakukan slicing untuk variabel dependen y, yaitu kolom terakhir yang menunjukkan status keluar/tidaknya pegawai.

In [5]:
# Mengonversi data kategori jenis kelamin menjadi numerik
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
from sklearn.compose import ColumnTransformer
labelencoder = LabelEncoder()
X[:, 2] = labelencoder.fit_transform(X[:, 2])

In [7]:
# Membuat dummy variable untuk kolom negara
columnTransformer = ColumnTransformer([('encoder', OneHotEncoder(), [1])], remainder='passthrough')
X = np.array(columnTransformer.fit_transform(X), dtype = np.float64)

# Menghilangkan 1 dummy variabel di kolom negara
X = X[:, 1:]

In [9]:
# Membagi data ke test dan training set
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)

# Feature Scaling
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

Membagi dataset ke dalam training dan test set.
Kemudian melakukan proses feature scaling. Hal ini adalah langkah wajib dalam deep learning, karena kita berhubungan dengan banyak sekali komputasi (beberapa kali epoch), sehingga kita tidak ingin skala satu kolom lebih mendominasi daripada skala kolom lainnya. 

In [10]:
# Mengimpor library Keras dan turunannya
# Sekarang menggunakan TensorFlow 2.0
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
# Inisialisasi ANN
MesinKlasifikasi = Sequential()
# Menambah input layer dan hidden layer pertama
MesinKlasifikasi.add(Dense(units = 6, kernel_initializer = 'uniform', activation = 'relu', input_dim = 11))
# Menambah hidden layer kedua
MesinKlasifikasi.add(Dense(units = 6, kernel_initializer = 'uniform', activation = 'relu'))
# Menambah output layer
MesinKlasifikasi.add(Dense(units = 1, kernel_initializer = 'uniform', activation = 'sigmoid'))
# Meenjalankan ANN
MesinKlasifikasi.compile(optimizer = 'adam', loss = 'binary_crossentropy', metrics = ['accuracy'])
# Fitting ANN ke training set
MesinKlasifikasi.fit(X_train, y_train, batch_size = 10, epochs = 100)

Epoch 1/100
800/800 [==============================] - 4s 3ms/step - loss: 0.4801 - accuracy: 0.7959
Epoch 2/100
800/800 [==============================] - 3s 4ms/step - loss: 0.4258 - accuracy: 0.7960
Epoch 3/100
800/800 [==============================] - 3s 4ms/step - loss: 0.4211 - accuracy: 0.8039
Epoch 4/100
800/800 [==============================] - 3s 3ms/step - loss: 0.4177 - accuracy: 0.8235
Epoch 5/100
800/800 [==============================] - 3s 3ms/step - loss: 0.4160 - accuracy: 0.8291
Epoch 6/100
800/800 [==============================] - 3s 3ms/step - loss: 0.4139 - accuracy: 0.8303
Epoch 7/100
800/800 [==============================] - 3s 3ms/step - loss: 0.4123 - accuracy: 0.8309
Epoch 8/100
800/800 [==============================] - 3s 4ms/step - loss: 0.4114 - accuracy: 0.8321
Epoch 9/100
800/800 [==============================] - 2s 3ms/step - loss: 0.4106 - accuracy: 0.8340
Epoch 10/100
800/800 [==============================] - 2s 3ms/step - loss: 0.4095 - accura

800/800 [==============================] - 2s 2ms/step - loss: 0.4001 - accuracy: 0.8347
Epoch 82/100
800/800 [==============================] - 2s 2ms/step - loss: 0.3999 - accuracy: 0.8363
Epoch 83/100
800/800 [==============================] - 2s 2ms/step - loss: 0.3999 - accuracy: 0.8353
Epoch 84/100
800/800 [==============================] - 2s 2ms/step - loss: 0.3995 - accuracy: 0.8360
Epoch 85/100
800/800 [==============================] - 2s 2ms/step - loss: 0.3998 - accuracy: 0.8354
Epoch 86/100
800/800 [==============================] - 2s 2ms/step - loss: 0.3997 - accuracy: 0.8351
Epoch 87/100
800/800 [==============================] - 2s 2ms/step - loss: 0.3995 - accuracy: 0.8356
Epoch 88/100
800/800 [==============================] - 2s 2ms/step - loss: 0.3997 - accuracy: 0.8354
Epoch 89/100
800/800 [==============================] - 2s 2ms/step - loss: 0.3993 - accuracy: 0.8346
Epoch 90/100
800/800 [==============================] - 2s 2ms/step - loss: 0.3998 - accuracy: 

- mengimpor modul dari tensorflow yang disebut dengan Sequential. Modul ini diperlukan untuk memulai neural networks. Kita mengimpornya dari sublibrary models pada keras, sehingga ditulisnya from tensorflow.keras.models import Sequential.
- mengimpor modul kedua dari keras yaitu Dense. Modul ini diperlukan untuk membuat layers di neural networks. Kita mengimpornya dari sublibrary layers di keras, jadi penulisannya adalah from tensorflow.keras.layers import Dense.
- mendefinisikan objek MesinKlasifikasi sebagai nama artificial neural networks kita. Tentu saja penamaan ini bebas. Caranya adalah cukup menjalankan fungsi (class) Sequential terhadap nama objek ini.
- menambah input layer (11 variabel independen) dan hidden layer (beberapa nodes di depan input layer). Caranya adalah menggunakan method add, kemudian diikuti dengan perintah Dense untuk menambahkan layernya. Di dalam perintah Dense, kita perlu mendefinisikan beberapa parameter.
- menambahkan hidden layer kedua, dengan jumlah neuron yang sama yaitu 6.
- membuat output layer. Mirip dengan line-line sebelumnya, kali ini kita memilih activation function yang berbeda yaitu sigmoid function, dan units=1 karena hanya ada 2 kategori (0/1).
-  menjalankan (meng-compile) neural network yang sudah kita desain dengan metode stochastic gradient descent.
- menjalankan (men-training) ANN dengan beberapa epoch yang kita tentukan, di mana setiap epoch ANN akan mengupdate nilai W dan terus berusaha meningkatkan akurasinya. Kita gunakan parameter batch_size=10, artinya kita lakukan forward dan back propagation setiap 10 baris data yang diproses. Kita tentukan jumlah epochs sebanyak 100 kali. 

In [11]:
# Memprediksi hasil test set
y_pred = MesinKlasifikasi.predict(X_test)
y_pred = (y_pred > 0.5)

63/63 [==============================] - 0s 3ms/step


In [13]:
# Membuat confusion matrix
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred)
print(cm)

[[1536   59]
 [ 251  154]]


Dapat lihat bahwa kita memiliki (1536+154 = 1690) prediksi benar dan (251+59 = 310) prediksi salah. Dengan demikian akurasinya adalah (1690/2000 = 0.845) 84.5%. Ternyata artificial neural networks (ANN) kita mampu memprediksi dengan cukup baik bahkan di atas 80%.

In [14]:
# evaluasi model
from sklearn.metrics import classification_report
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.86      0.96      0.91      1595
           1       0.72      0.38      0.50       405

    accuracy                           0.84      2000
   macro avg       0.79      0.67      0.70      2000
weighted avg       0.83      0.84      0.83      2000

